src/components/Table/EditColumnsButton.tsx

```diff
@@ -0,0 +1,89 @@
+import { Dispatch, SetStateAction, useCallback, useEffect, useMemo, useRef, useState } from "react";
+import { useMenuTrigger } from "react-aria";
+import { useMenuTriggerState } from "react-stately";
+import { CheckboxGroup, CheckboxGroupItemOption } from "src/inputs";
+import { Css } from "../../Css";
+import { useTestIds } from "../../utils";
+import { Button } from "../Button";
+import { isIconButton, isTextButton, OverlayTrigger, OverlayTriggerProps } from "../internal/OverlayTrigger";
+import { GridColumn, Kinded } from "./GridTable";
+
+interface EditColumnsButtonProps<R extends Kinded, S>
+  extends Pick<OverlayTriggerProps, "trigger" | "placement" | "disabled" | "tooltip"> {
+  columns: GridColumn<R, S>[];
+  setColumns: Dispatch<SetStateAction<GridColumn<R, S>[]>>;
+  title?: string;
+  // for storybook purposes
+  defaultOpen?: boolean;
+}
+
+export function EditColumnsButton<R extends Kinded, S = {}>(props: EditColumnsButtonProps<R, S>) {
+  const { defaultOpen, disabled, columns, setColumns, trigger, title } = props;
+  const state = useMenuTriggerState({ isOpen: defaultOpen });
+  const buttonRef = useRef<HTMLButtonElement>(null);
+  const { menuTriggerProps } = useMenuTrigger({ isDisabled: !!disabled }, state, buttonRef);
+  const tid = useTestIds(
+    props,
+    isTextButton(trigger) ? trigger.label : isIconButton(trigger) ? trigger.icon : trigger.name,
+  );
+
+  const { options, selectedColumns } = useMemo(() => {
+    const editableColumns = columns.filter((column) => {
+      if (!column.canHide) return false;
+      if (!column.name || column.name.length === 0) {
+        console.warn("Column is missing 'name' property required by the Edit Columns button", column);
+        return false;
+      }
+      return true;
+    });
+
+    const selectedColumns = editableColumns
+      .map((column) => {
+        if (column.canHide && column.visible) return column.name;
+      })
+      .filter((column) => !!column) as string[];
+
+    const options: CheckboxGroupItemOption[] = editableColumns.map((column) => ({
+      label: column.name!,
+      value: column.name!,
+    }));
+
+    return { selectedColumns, options };
```

Comment:
**thought (non-blocking):** Personally, I like to see how few times I need to loop in order to get the values I need. In this instance you could use something like a `.reduce` method, which could achieve the `console.warn` and properly group the `selectedColumns`, and `options` values. I don't think this is a problem the way you have it, its a very small component, which will typically take a very small data set, so doing this loop a few times isn't going to hurt us.
Just because I like to explore these sort of things (they remind me of interview questions), then I'm going to take a stab at what that might look like:
```
const { options, selectedColumns } = useMemo(() => {
  return columns.reduce((acc, col) => {
    // Only include options that can be hidden and have the `name` property defined.
    if (!column.canHide) return acc;
    if (!column.name || column.name.length === 0) {
      console.warn("Column is missing 'name' property required by the Edit Columns button", column);
      return acc;
    }
    
    // If currently visible, then add to selectedColumns
    if (column.canHide && column.visible) {
      acc.selectedColumns.push(column);
    }
    
    // Add current column as an option
    return { ...acc, options: acc.options.concat({ label: column.name!, value: column.name! }) };
  }, { options: [] as CheckboxGroupItemOption[], selectedColumns as string[] }); 
```